lib/expr.ml

```diff
@@ -0,0 +1,280 @@
+open Ty
+
+type t =
+  { e : expr
+  ; ty : Ty.t
+  }
+
+and expr =
+  | Val of Value.t
+  | Ptr of int32 * t
+  | Unop of unop * t
+  | Binop of binop * t * t
+  | Triop of triop * t * t * t
+  | Relop of relop * t * t
+  | Cvtop of cvtop * t
+  | Symbol of Symbol.t
+  | Extract of t * int * int
+  | Concat of t * t
+
+let ( @: ) e ty = { e; ty }
+let mk_symbol s = Symbol s @: Symbol.type_of s
+let is_num e = match e.e with Val (Num _) -> true | _ -> false
+
+let rec equal (e1 : t) (e2 : t) : bool =
+  if not (e1.ty = e2.ty) then false
+  else
+    match (e1.e, e2.e) with
+    | Val v1, Val v2 -> Value.equal v1 v2
+    | Ptr (b1, o1), Ptr (b2, o2) -> b1 = b2 && equal o1 o2
+    | Unop (op1, e1), Unop (op2, e2) -> op1 = op2 && equal e1 e2
+    | Cvtop (op1, e1), Cvtop (op2, e2) -> op1 = op2 && equal e1 e2
+    | Binop (op1, e1, e3), Binop (op2, e2, e4) ->
+      op1 = op2 && equal e1 e2 && equal e3 e4
+    | Relop (op1, e1, e3), Relop (op2, e2, e4) ->
+      op1 = op2 && equal e1 e2 && equal e3 e4
+    | Triop (op1, e1, e3, e5), Triop (op2, e2, e4, e6) ->
+      op1 = op2 && equal e1 e2 && equal e3 e4 && equal e5 e6
+    | Symbol s1, Symbol s2 -> Symbol.equal s1 s2
+    | Extract (e1, h1, l1), Extract (e2, h2, l2) ->
+      equal e1 e2 && h1 = h2 && l1 = l2
+    | Concat (e1, e3), Concat (e2, e4) -> equal e1 e2 && equal e3 e4
+    | _ -> false
+
+let get_symbols (e : t list) : Symbol.t list =
+  let rec symbols e =
+    match e.e with
+    | Val _ -> []
+    | Ptr (_, offset) -> symbols offset
+    | Unop (_, e1) -> symbols e1
+    | Binop (_, e1, e2) -> symbols e1 @ symbols e2
+    | Triop (_, e1, e2, e3) -> symbols e1 @ symbols e2 @ symbols e3
+    | Relop (_, e1, e2) -> symbols e1 @ symbols e2
+    | Cvtop (_, e) -> symbols e
+    | Symbol s -> [ s ]
+    | Extract (e, _, _) -> symbols e
+    | Concat (e1, e2) -> symbols e1 @ symbols e2
+  in
+  List.fold_left
+    (fun accum x -> if List.mem x accum then accum else x :: accum)
+    []
+    (List.concat_map symbols e)
```

Comment:
The complexity of this function can be made much lower: in symbols, you could avoid concatenating lists which is expensive, in the `fold_left` call, `List.mem` on every element is making the whole thing quadratic. I would suggest something like:

```ocaml
let get_symbols e =
  let tbl = Hashtbl.create 64 in
  let rec symbols e = 
    match e.e with
    | Val _ -> ()
    | Ptr (_, offset) -> symbols offset
    | Unop (_, e1) -> symbols e1
    | Binop (_, e1, e2) -> symbols e1; symbols e2
    | Triop (_, e1, e2, e3) -> symbols e1; symbols e2; symbols e3
    | Relop (_, e1, e2) -> symbols e1; symbols e2
    | Cvtop (_, e) -> symbols e
    | Symbol s -> Hashtbl.replace tbl s ()
    | Extract (e, _, _) -> symbols e
    | Concat (e1, e2) -> symbols e1; symbols e2
  in
  List.iter symbols e;
  Hashtbl.fold (fun k () acc -> k::acc) tbl []
``` 
  (untested)

Reply:
Ah yes! Been meaning to optimize this function, this is perfect thanks!! :smiley: